improving the advanced indexing notebooks

negin513 opened the pull request:
This PR should address issue #206 . 

 I created and added a schematic image to explain different indexing types. I also added more examples to the notebook for more clarify.

negin513 commented:
I am not sure why the preview of notebook is failing. There is a duplicate label warning : 
```
 WARNING: duplicate label: indexing-1; other instance in /home/runner/work/xarray-tutorial/xarray-tutorial/fundamentals/02.1_indexing_Basic.ipynb
 ```

But I don't see any similar titles.

dcherian commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>Would be good to expand this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

negin513 replied:
<p>Done! The updated learning objectives:</p><p> </p><p>## Learning Objectives</p><p> </p><p>* Orthogonal vs. Pointwise (Vectorized) Indexing.</p><p>* Pointwise indexing in Xarray to extract data at a collection of points.</p><p>* Understand the difference between NumPy and Xarray indexing behavior.</p>

dcherian commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>The text is good but might be hard to understand to a reader with no context.</p><p> </p><p>Perhaps move the example image from below up above these descriptions. So then there's a visual context that a reader can refer back to.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

JessicaS11 replied:
<p>I second this suggestion!</p><p> </p>

negin513 replied:
Thanks @dcherian and @JessicaS11 for both of your suggestions on this. I had a hard time with the flow of this notebook and after implementing your comments, I think the structure is much clearer and more logical.

negin513 replied:
<p>So now, it is structured as following. </p><p> </p><ul><li>Examples of indexing on 5x5 2d arrays + the diagram</li><li>Explanation of Pointwise Indexing and Orthogonal Indexing. </li></ul><p> </p><p> </p>

dcherian commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>Consider moving the formal definitions up top down here.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

negin513 replied:
Done! Thanks for the comment.

dcherian commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>typo in Exercises.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

JessicaS11 replied:
<p>extra space before the colon (before {solution})</p>

negin513 replied:
Should be fixed now! Thanks for noticing these.

dcherian commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>Could also link to <a href="https://jakevdp.github.io/PythonDataScienceHandbook/02.07-fancy-indexing.html" target="_blank">https://jakevdp.github.io/PythonDataScienceHandbook/02.07-fancy-indexing.html</a></p><p> </p><p>and stick it in a <code>seealso</code> block </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

negin513 replied:
Thanks for suggesting this. I added this and NumPy advanced indexing (https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing) in the seealso block.

dcherian commented:
👏🏾 :clap: i love this image! 

we should get it in the main docs too.

JessicaS11 commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>make "point-wise indexing" bold</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

negin513 replied:
Done!

JessicaS11 commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>As a newbie to advanced indexing: it seems to me that in both orthogonal and vectorized indexing a "different name than the original dimensions" needs to be specified. Instead, the critical part for triggering vectorized indexing is that ALL selections need to target the SAME selection dimension. So for orthogonal, we create two dimensions <code>degrees_north</code> and <code>degrees_east</code>, while in vectorized we create a single <code>points</code> dimension.</p><p> </p><p>Assuming my understanding is correct, how about: "To trigger vectorized..., you will need to provide the selection dimensions with a single shared output dimension name that is different than each of the original dimensions. This dimension name...". I fully appreciate how difficult this is to articulate clearly!</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

negin513 replied:
@JessicaS11 You are correct! Thanks so much for your suggestions. My original wordings was confusing. Using your suggestion, I updated this sentence as following:

To trigger vectorized indexing behavior, you will need to provide the selection dimensions with a new **shared** output dimension name. This means that the dimensions of both indexers must be the same, and the output will have the same dimension name as the indexers.

I made sure to emphasize on the *same* dimension name in a few places.

JessicaS11 commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>Conceptually, would it make more sense here if we used <code>lon=target_lon</code> because we'd most likely want that to be on the <code>degrees_east</code> dim?  </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

negin513 replied:
Yes and I updated the notebook to reflect this.

dcherian commented on intermediate/indexing/advanced-indexing.ipynb:
```diff
@@ -8,7 +8,7 @@
     "\n",
```
<p>I would note that there are no shared dimension names between <code>target_lat</code> and <code>target_lon</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/xarray-contrib/xarray-tutorial/pull/264/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

negin513 replied:
Good point! I updated this as following:

In the above example, you can see how the output shape is `time` x `lats` x `lons`. Please note that there are no shared dimensions between the indexers, so the output shape is the union of the dimensions of the indexers.

scottyhq commented:
@negin513 just a heads up - I added a couple commits to sync with the latest environment and enable the website preview link, so you can see your changes here https://xarray-contrib-xarray-tutorial-preview-pr-264.surge.sh/intermediate/indexing/advanced-indexing.html

scottyhq reviewed (approved):
This is a wonderful improvement @negin513 thank you! I love the image and narrative flows really well. Feel free to merge. two small ideas:

1. The plot will look better on the website if you add `%config InlineBackend.figure_format='retina'` in the top cell
2. Not sure if it's worth adding a sentence or linking to [np.ix_](https://numpy.org/doc/stable/reference/generated/numpy.ix_.html#numpy.ix_) to achieve the orthogonal result in numpy `np_array[np.ix_([0, 2, 4], [0, 2, 4])]` If someone knows why this is named so obscurely (.ix_) I'd love to know :)

dcherian commented:
I think meshgrid is an alternative to ix_. I won't have time to review but thanks Negin!

negin513 commented:
Thanks @scottyhq for great comments . I have added  the following: 
1. Added a dropdown note box about how to achieve numpy orthogonal indexing. 
2.  added `%config InlineBackend.figure_format='retina'` for rendering
